### PR TITLE
Add majority of datasource functionality

### PR DESCRIPTION
**Pull request by ashulmanWeWork**

### CODE REVIEW

File: docs/openapi.yml [hunk 1]
```diff
@@ -336,7 +336,7 @@ components:
           type: string
           format: date-time
         owner:
-          description: The owner of the namespace.
+          description: The connectionUrl of the namespace.
```
Comment (wslulciuc) [inline docs/openapi.yml]: You'll want to be careful with global renames. Not sure this is what you intended?

File: src/main/java/marquez/api/mappers/DatasourceResponseMapper.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  private DatasourceResponseMapper() {}
+
+  public static DatasourceResponse map(@NonNull Datasource datasource) {
+    return new DatasourceResponse(
+        datasource.getCreatedAt().toString(),
+        datasource.getDataSourceName().getValue(),
+        datasource.getConnectionUrl().getRawValue());
+  }
+
+  public static List<DatasourceResponse> map(@NonNull List<Datasource> datasources) {
```
Comment (wslulciuc) [inline src/main/java/marquez/api/mappers/DatasourceResponseMapper.java]: You'll want to follow the recommended approach to return a list of responses: https://github.com/MarquezProject/marquez/issues/309

File: src/main/java/marquez/api/models/NamespaceResponse.java [hunk 3]
```diff
@@ -30,7 +30,7 @@
   @JsonProperty("createdAt")
   String createdAt;
 
-  @JsonProperty("owner")
+  @JsonProperty("connectionUrl")
```
Comment (wslulciuc) [inline src/main/java/marquez/api/models/NamespaceResponse.java]: Did you mean to rename this field?

File: src/main/java/marquez/common/models/ConnectionUrl.java [hunk 4]
```diff
@@ -22,12 +22,12 @@
 @EqualsAndHashCode
 @ToString
 public final class ConnectionUrl {
-  @Getter private final DataSource dataSource;
```
Comment (wslulciuc) [inline src/main/java/marquez/common/models/ConnectionUrl.java]: We should renamed the class `DataSource` in seperate PRs. Going forward, we'll need to have an issue. Mind still crating one for the name change and link it in your PR?

File: src/main/java/marquez/service/DatasourceService.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    final Optional<DatasourceRow> datasourceRowIfFound = dataSourceDao.findBy(datasourceUUID);
+    try {
+      return datasourceRowIfFound.map(DatasourceMapper::map).orElseThrow(UnexpectedException::new);
+    } catch (UnexpectedException e) {
+      log.error(e.getMessage());
+      throw new UnexpectedException();
+    }
+  }
+
+  public List<Datasource> list(@NonNull Integer limit, @NonNull Integer offset) {
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: We'll want to remain consistent across services when retrieving a list of `X`. So, I recommend going with `DatasourceService.getAll()`

see: https://github.com/MarquezProject/marquez/blob/master/src/main/java/marquez/service/DatasetService.java#L76

File: src/test/java/marquez/db/models/OwnerRowTest.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
 public class OwnerRowTest {
   private static final UUID ROW_UUID = UUID.randomUUID();
   private static final Instant CREATED_AT = Instant.now();
-  private static final String NAME = "test owner";
+  private static final String NAME = "test connectionUrl";
```
Comment (wslulciuc) [inline src/test/java/marquez/db/models/OwnerRowTest.java]: Same comment around renaming

File: src/test/java/marquez/db/models/NamespaceRowTest.java [hunk 7]
```diff
@@ -26,7 +26,7 @@
   private static final Instant CREATED_AT = Instant.now();
   private static final Instant UPDATED_AT = Instant.now();
   private static final String NAME = "test_namespace";
-  private static final String CURRENT_OWNER_NAME = "test owner";
+  private static final String CURRENT_OWNER_NAME = "test connectionUrl";
```
Comment (wslulciuc) [inline src/test/java/marquez/db/models/NamespaceRowTest.java]: Same comment around renaming

File: src/main/java/marquez/api/resources/DatasourceResource.java [hunk 8]
```diff
@@ -28,6 +34,9 @@ public DatasourceResource() {}
   public Response list(
       @QueryParam("limit") @DefaultValue("100") Integer limit,
       @QueryParam("offset") @DefaultValue("0") Integer offset) {
-    return Response.ok(new DatasourcesResponse(Collections.emptyList())).build();
+
+    List<marquez.service.models.Datasource> datasourceList = datasourceService.list(limit, offset);
```
Comment (wslulciuc) [inline src/main/java/marquez/api/resources/DatasourceResource.java]: Is there a reason we are specifying the fully-qualified path for `Datasource`?

File: src/main/java/marquez/api/resources/DatasourceResource.java [hunk 9]
```diff
@@ -5,19 +5,25 @@
...
 import marquez.api.models.DatasourcesResponse;
+import marquez.service.DatasourceService;
 
 @Path("/api/v1")
 public final class DatasourceResource {
 
-  public DatasourceResource() {}
+  private final DatasourceService datasourceService;
+
+  public DatasourceResource(DatasourceService dataSourceService) {
```
Comment (wslulciuc) [inline src/main/java/marquez/api/resources/DatasourceResource.java]: minor: by convention, we'll want to define the constructor as:

```java
public DatasourceResource(final DatasourceService datasourceService) {
  ...
}
```
That is,
*  Args passed to the construct are always `final`
* `dataSourceService` **=>** `datasourceService`

File: src/main/java/marquez/api/resources/DatasourceResource.java (hunk 8 above)
Comment (wslulciuc) [reply src/main/java/marquez/api/resources/DatasourceResource.java]: minor: we know that `datasourceList` is a `List`, therefore the suffix is redundant, meaning we can just take the plural form: `datasources`

File: src/main/java/marquez/service/models/Datasource.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+import marquez.common.models.ConnectionUrl;
+import marquez.common.models.DatasourceName;
+
+@RequiredArgsConstructor
+@EqualsAndHashCode
+@ToString
+public class Datasource {
+
```
Comment (wslulciuc) [inline src/main/java/marquez/service/models/Datasource.java]: minor: remove white space :)

File: src/main/java/marquez/service/models/Datasource.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  @Getter @NonNull private final Instant createdAt;
+  @Getter @NonNull private final DatasourceName dataSourceName;
```
Comment (wslulciuc) [inline src/main/java/marquez/service/models/Datasource.java]: minor: it's great that we defined the `DatasourceName` type, but when used in `Datasource`, let's name `dataSourceName` as just `name`: meaning `Datasource.name` reads better than `Datasource.dataSourceName`

File: src/main/java/marquez/common/models/DatasourceName.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package marquez.common.models;
```
Comment (wslulciuc) [inline src/main/java/marquez/common/models/DatasourceName.java]: minor: add license

File: src/main/java/marquez/service/DatasourceService.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+package marquez.service;
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: minor: add license

File: src/main/java/marquez/service/DatasourceService.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+  public DatasourceService(@NonNull final DatasourceDao dataSourceDao) {
+    this.dataSourceDao = dataSourceDao;
+  }
+
+  public Datasource create(@NonNull ConnectionUrl connectionUrl, @NonNull DatasourceName name)
+      throws UnexpectedException {
+    UUID datasourceUUID = UUID.randomUUID();
+    dataSourceDao.insert(
+        new DatasourceRow(datasourceUUID, name.getValue(), connectionUrl.getRawValue(), null));
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: You'll need to define a `DatasourceRowMapper` to instantiate **class** `DatasourceRow`, use `DatasetRowMapper` as an example: https://github.com/MarquezProject/marquez/blob/master/src/main/java/marquez/service/mappers/DatasetMapper.java

File: src/main/java/marquez/service/DatasourceService.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    final Optional<DatasourceRow> datasourceRowIfFound = dataSourceDao.findBy(datasourceUUID);
+    try {
+      return datasourceRowIfFound.map(DatasourceMapper::map).orElseThrow(UnexpectedException::new);
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: The **class** `UnexpectedException` was removed in #340, mind making sure you're working with the latest changes in `master`?

File: src/main/java/marquez/service/mappers/DatasetRowMapper.java [hunk 16]
```diff
@@ -19,16 +19,16 @@
...
+import marquez.db.models.DatasourceRow;
 import marquez.service.models.DbTableVersion;
 
 public final class DatasetRowMapper {
   private DatasetRowMapper() {}
 
   public static DatasetRow map(
       @NonNull NamespaceName namespaceName,
-      @NonNull DataSourceRow dataSourceRow,
+      @NonNull DatasourceRow dataSourceRow,
```
Comment (wslulciuc) [inline src/main/java/marquez/service/mappers/DatasetRowMapper.java]: minor: shouldn't we be consistent in how we are naming datasource?

File: src/main/java/marquez/service/DatasourceService.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  public DatasourceService(@NonNull final DatasourceDao dataSourceDao) {
+    this.dataSourceDao = dataSourceDao;
+  }
+
+  public Datasource create(@NonNull ConnectionUrl connectionUrl, @NonNull DatasourceName name)
+      throws UnexpectedException {
+    UUID datasourceUUID = UUID.randomUUID();
+    dataSourceDao.insert(
+        new DatasourceRow(datasourceUUID, name.getValue(), connectionUrl.getRawValue(), null));
+    final Optional<DatasourceRow> datasourceRowIfFound = dataSourceDao.findBy(datasourceUUID);
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: Our style check will warn about 2 subsequent uppercase chars: `datasourceUUID` **=>** `datasourceUuid`

File: src/main/java/marquez/service/DatasourceService.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+@Slf4j
+public class DatasourceService {
+  private final DatasourceDao dataSourceDao;
+
+  public DatasourceService(@NonNull final DatasourceDao dataSourceDao) {
+    this.dataSourceDao = dataSourceDao;
+  }
+
+  public Datasource create(@NonNull ConnectionUrl connectionUrl, @NonNull DatasourceName name)
```
Comment (wslulciuc) [inline src/main/java/marquez/service/DatasourceService.java]: Are you sure `Datasource. create()` is creating a datasource?

File: docs/openapi.yml (hunk 1 above)
Comment (ashulmanWeWork) [reply docs/openapi.yml]: Good call - sorry. Will be fixed in next revision.

File: src/main/java/marquez/api/mappers/DatasourceResponseMapper.java (hunk 2 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/mappers/DatasourceResponseMapper.java]: Nice - much cleaner that way. Will be in next revision.

File: src/main/java/marquez/api/models/NamespaceResponse.java (hunk 3 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/models/NamespaceResponse.java]: Will be fixed in the next revision.

File: src/main/java/marquez/common/models/ConnectionUrl.java (hunk 4 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/common/models/ConnectionUrl.java]: Sure - filed #360.

Comment (ashulmanWeWork) [reply src/main/java/marquez/common/models/ConnectionUrl.java]: We've settled on the class naming to be `Datasource` - should the variable naming also be `datasource`?

File: src/main/java/marquez/service/DatasourceService.java (hunk 5 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Sounds good. Will be fixed in next revision.

File: src/main/java/marquez/api/resources/DatasourceResource.java (hunk 8 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/resources/DatasourceResource.java]: Good catch. No reason for that to be fully-qualified. Will also rename to datasources. This'll be fixed in the next revision.

File: src/main/java/marquez/api/resources/DatasourceResource.java (hunk 9 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/resources/DatasourceResource.java]: Will be fixed in the next revision.

File: docs/openapi.yml (hunk 1 above)
Comment (sshah-wework) [reply docs/openapi.yml]: what is the next revision? can it be fixed here?

File: src/main/java/marquez/service/DatasourceService.java (hunk 13 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: I'll add the license.
We should consider making this a build rule. Spotless has a way of helping with this:
https://github.com/diffplug/spotless/tree/master/plugin-gradle#license-header-options

File: docs/openapi.yml (hunk 1 above)
Comment (ashulmanWeWork) [reply docs/openapi.yml]: Yup - the next version of this patch I upload will have the fix.

File: src/main/java/marquez/service/models/Datasource.java (hunk 10 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/models/Datasource.java]: Will be fixed in next revision.

File: src/main/java/marquez/service/models/Datasource.java (hunk 11 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/models/Datasource.java]: Sounds good. Will take care of this in the next revision.

File: src/main/java/marquez/service/DatasourceService.java (hunk 14 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: That'll def make it cleaner. Will add.

File: src/main/java/marquez/common/models/ConnectionUrl.java (hunk 4 above)
Comment (wslulciuc) [reply src/main/java/marquez/common/models/ConnectionUrl.java]: Yes, for consistency

File: src/test/java/marquez/db/models/NamespaceRowTest.java (hunk 7 above)
Comment (sshah-wework) [reply src/test/java/marquez/db/models/NamespaceRowTest.java]: :( so many of these

File: src/test/java/marquez/service/models/Generator.java [hunk 19]
```diff
@@ -156,13 +160,23 @@ public static RunArgs cloneRunArgs(RunArgs ra) {
     return new RunArgs(ra.getHexDigest(), ra.getJson(), ra.getCreatedAt());
   }
 
+  // Datasource
+  public static Datasource genDatasource() {
+    int random = (int) (Math.random() * 50 + 1);
```
Comment (sshah-wework) [inline src/test/java/marquez/service/models/Generator.java]: Two questions
* there is already a `randNum` in Generator, why not use that?
* why are we multiplying by 50 and adding 1?

File: src/test/java/marquez/service/models/Generator.java [hunk 20]
```diff
@@ -156,13 +160,23 @@ public static RunArgs cloneRunArgs(RunArgs ra) {
...
+    int random2 = (int) (Math.random() * 50 + 1);
```
Comment (sshah-wework) [inline src/test/java/marquez/service/models/Generator.java]: what is `random2` used for?

**Comment by sshah-wework** [comment]
Also, just an FYI, try to put the "Fixes #..." should in the PR description and not the title to keep things tidy

**Comment by wslulciuc** [comment]
Good point @sshah-wework, we also outline how to link issues in PRs in our [CONTRIBUTING.md](https://github.com/MarquezProject/marquez/blob/master/CONTRIBUTING.md)

/cc @ashulmanWeWork

File: src/main/java/marquez/service/DatasourceService.java [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+import marquez.db.models.DatasourceRow;
+import marquez.service.exceptions.UnexpectedException;
+import marquez.service.mappers.DatasourceMapper;
+import marquez.service.models.Datasource;
+
+@Slf4j
+public class DatasourceService {
+  private final DatasourceDao dataSourceDao;
+
+  public DatasourceService(@NonNull final DatasourceDao dataSourceDao) {
```
Comment (sshah-wework) [inline src/main/java/marquez/service/DatasourceService.java]: what about `get()` for looking up a single `Datasource`?

**Comment by ashulmanWeWork** [comment]
Fixes #352 (moving from title)

**Comment by ashulmanWeWork** [comment]
Need to add validation around connectionUrls, as well as validation around name:type uniqueness.

File: docs/openapi.yml (hunk 1 above)
Comment (ashulmanWeWork) [reply docs/openapi.yml]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/api/mappers/DatasourceResponseMapper.java (hunk 2 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/mappers/DatasourceResponseMapper.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/api/models/NamespaceResponse.java (hunk 3 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/models/NamespaceResponse.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/common/models/ConnectionUrl.java (hunk 4 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/common/models/ConnectionUrl.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 5 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/test/java/marquez/db/models/OwnerRowTest.java (hunk 6 above)
Comment (ashulmanWeWork) [reply src/test/java/marquez/db/models/OwnerRowTest.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/test/java/marquez/db/models/NamespaceRowTest.java (hunk 7 above)
Comment (ashulmanWeWork) [reply src/test/java/marquez/db/models/NamespaceRowTest.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/api/resources/DatasourceResource.java (hunk 8 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/resources/DatasourceResource.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/api/resources/DatasourceResource.java (hunk 9 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/api/resources/DatasourceResource.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/models/Datasource.java (hunk 10 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/models/Datasource.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/models/Datasource.java (hunk 11 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/models/Datasource.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/common/models/DatasourceName.java (hunk 12 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/common/models/DatasourceName.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 13 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 14 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 15 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/mappers/DatasetRowMapper.java (hunk 16 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/mappers/DatasetRowMapper.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 17 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 18 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Adjusted what was in the try block. Now resolved in PRs #394, #387, and #384, where this is fixed.

File: src/test/java/marquez/service/models/Generator.java (hunk 19 above)
Comment (ashulmanWeWork) [reply src/test/java/marquez/service/models/Generator.java]: Please see PRs #394, #387, and #384, where this is fixed.

File: src/test/java/marquez/service/models/Generator.java (hunk 20 above)
Comment (ashulmanWeWork) [reply src/test/java/marquez/service/models/Generator.java]: Good catch. Removed in PRs #394, #387, and #384, where this is fixed.

File: src/main/java/marquez/service/DatasourceService.java (hunk 21 above)
Comment (ashulmanWeWork) [reply src/main/java/marquez/service/DatasourceService.java]: Please see PRs #394, #387, and #384, where this is fixed.

**Pull request closed without merging**